File: aural/process-reddit-url.js
#!/usr/bin/env node

/**
 * Process Reddit URL - Complete End-to-End Workflow
 * 
 * Takes a Reddit URL as input and:
 * 1. Fetches the Reddit post
 * 2. Analyzes it with LLM
 * 3. Extracts all audio files
 * 4. Creates a complete release
 */

const { execSync } = require('child_process');
const fs = require('fs').promises;
const path = require('path');
const { ReleaseOrchestrator } = require('./release-orchestrator');

class RedditProcessor {
  constructor(config = {}) {
    this.outputDir = config.outputDir || 'data';
    this.tempDir = path.join(this.outputDir, '.temp');
    this.orchestrator = new ReleaseOrchestrator({
      dataDir: this.outputDir,
      cacheEnabled: true,
      validateExtractions: true
    });
  }

  /**
   * Process a Reddit URL through the complete pipeline
   */
  async processRedditUrl(redditUrl) {
    console.log('🚀 Processing Reddit URL:', redditUrl);
    console.log('=' .repeat(60));
    
    try {
      // Ensure temp directory exists
      await fs.mkdir(this.tempDir, { recursive: true });
      
      // Step 1: Analyze Reddit post
      console.log('\n📝 Step 1: Analyzing Reddit post...');
      const analysis = await this.analyzeRedditPost(redditUrl);
      console.log(`✅ Analysis complete: ${analysis.metadata.title}`);
      
      // Step 2: Extract Reddit post data
      console.log('\n📊 Step 2: Extracting post metadata...');
      const redditPost = this.extractRedditData(analysis);
      console.log(`✅ Found ${analysis.audio_versions?.length || 0} audio version(s)`);
      
      // Step 3: Process into release
      console.log('\n🎵 Step 3: Extracting audio files...');
      const release = await this.orchestrator.processPost(redditPost, analysis);
      
      // Step 4: Generate summary
      console.log('\n📋 Step 4: Generating summary...');
      const summary = await this.generateSummary(release, redditUrl);
      
      // Cleanup temp directory
      try {
        await fs.rm(this.tempDir, { recursive: true });
      } catch (e) {
        // Ignore cleanup errors
      }
      
      return {
        success: true,
        release,
        summary
      };
      
    } catch (error) {
      console.error('\n❌ Processing failed:', error.message);
      throw error;
    }
  }

  /**
   * Analyze Reddit post using the existing analyze-reddit-post.js
   */
  async analyzeRedditPost(redditUrl) {
    const analysisOutput = path.join(this.tempDir, 'analysis.json');
    
    try {
      // Run the Reddit post analyzer
      const command = `node analyze-reddit-post.js "${redditUrl}" --output "${analysisOutput}"`;
      console.log('  Running:', command);
      
      const result = execSync(command, {
        encoding: 'utf8',
        stdio: 'pipe',
        cwd: __dirname
      });
      
      // Read the analysis result
      const analysisData = await fs.readFile(analysisOutput, 'utf8');
      return JSON.parse(analysisData);
      
    } catch (error) {
      // If analyze-reddit-post.js doesn't exist or fails, create a basic analysis
      console.warn('⚠️  Could not run analyze-reddit-post.js, using basic extraction');
      return await this.basicRedditAnalysis(redditUrl);
    }
  }

  /**
   * Basic Reddit analysis fallback
   */
  async basicRedditAnalysis(redditUrl) {
    // Extract post ID from URL
    const postIdMatch = redditUrl.match(/comments\/([a-z0-9]+)/);
    if (!postIdMatch) {
      throw new Error('Invalid Reddit URL format');
    }
    
    const postId = postIdMatch[1];
    
    // First, check if we have local Reddit data
    const localDataPaths = [
      path.join('reddit_data', '*', `${postId}_*.json`),
      path.join('data', 'reddit', '*', `${postId}_*.json`),
      path.join('H:', 'Git', 'gwasi-extractor', 'reddit_data', '*', `${postId}_*.json`)
    ];
    
    let post = null;
    
    for (const pattern of localDataPaths) {
      try {
        const glob = require('glob');
        const files = glob.sync(pattern);
        if (files.length > 0) {
          console.log('  Found local Reddit data:', files[0]);
          const localData = JSON.parse(await fs.readFile(files[0], 'utf8'));
          post = localData.reddit_data || localData;
          break;
        }
      } catch (e) {
        // Continue trying other paths
      }
    }
    
    // If no local data, try Reddit API
    if (!post) {
      const jsonUrl = redditUrl.replace(/\/$/, '') + '.json';
      console.log('  Fetching:', jsonUrl);
      
      const response = await fetch(jsonUrl, {
        headers: {
          'User-Agent': 'GWASIExtractor/1.0'
        }
      });
      
      if (!response.ok) {
        throw new Error(`Failed to fetch Reddit post: ${response.status}`);
      }
      
      const data = await response.json();
      post = data[0].data.children[0].data;
    }
    
    // Extract audio URLs from post content
    const content = post.selftext || '';
    const audioUrls = this.extractAudioUrls(content);
    
    // Build analysis structure
    return {
      metadata: {
        post_id: post.id,
        title: post.title,
        username: post.author,
        created_utc: new Date(post.created_utc * 1000).toISOString(),
        url: redditUrl
      },
      content: {
        title: post.title,
        selftext: content
      },
      audio_versions: audioUrls.length > 0 ? [{
        version_name: 'Main',
        urls: audioUrls.map(url => ({
          url: url,
          platform: this.detectPlatform(url)
        }))
      }] : [],
      tags: this.extractTags(post.title + ' ' + content)
    };
  }

  /**
   * Extract audio URLs from text
   */
  extractAudioUrls(text) {
    const urls = [];
    const patterns = [
      /https?:\/\/(?:www\.)?soundgasm\.net\/u\/[^\s\]\)]+/gi,
      /https?:\/\/(?:www\.)?whyp\.it\/tracks\/[^\s\]\)]+/gi,
      /https?:\/\/(?:www\.)?hotaudio\.net\/u\/[^\s\]\)]+/gi
    ];
    
    for (const pattern of patterns) {
      const matches = text.match(pattern) || [];
      // Clean up URLs that might have trailing parentheses
      const cleanedUrls = matches.map(url => {
        // Remove trailing ) if it's not part of the URL structure
        if (url.endsWith(')') && !url.includes('(')) {
          return url.slice(0, -1);
        }
        return url;
      });
      urls.push(...cleanedUrls);
    }
    
    return [...new Set(urls)]; // Remove duplicates
  }

  /**
   * Detect platform from URL
   */
  detectPlatform(url) {
    if (/soundgasm\.net/i.test(url)) return 'Soundgasm';
    if (/whyp\.it/i.test(url)) return 'Whyp.it';
    if (/hotaudio\.net/i.test(url)) return 'HotAudio';
    return 'Unknown';
  }

  /**
   * Extract tags from text
   */
  extractTags(text) {
    const tags = [];
    const tagRegex = /\[([^\]]+)\]/g;
    let match;
    
    while ((match = tagRegex.exec(text)) !== null) {
      tags.push(match[1]);
    }
    
    return tags;
  }

  /**
   * Convert analysis to Reddit post format for orchestrator
   */
  extractRedditData(analysis) {
    return {
      id: analysis.metadata.post_id,
      title: analysis.metadata.title,
      author: analysis.metadata.username,
      created_utc: analysis.metadata.created_utc,
      selftext: analysis.content?.selftext || '',
      url: analysis.metadata.url
    };
  }

  /**
   * Generate summary report
   */
  async generateSummary(release, redditUrl) {
    const summary = {
      reddit_url: redditUrl,
      release_id: release.id,
      title: release.title,
      performer: release.primaryPerformer,
      audio_sources: release.audioSources.length,
      platforms: [...new Set(release.audioSources.map(s => s.metadata.platform.name))],
      extracted_files: [],
      timestamp: new Date().toISOString()
    };
    
    // List extracted files
    for (const source of release.audioSources) {
      summary.extracted_files.push({
        platform: source.metadata.platform.name,
        file: source.audio.filePath,
        size: source.audio.fileSize,
        checksum: source.audio.checksum?.sha256
      });
    }
    
    // Save summary
    const summaryPath = path.join(
      this.outputDir,
      'releases',
      release.primaryPerformer,
      release.id,
      'summary.json'
    );
    
    await fs.writeFile(summaryPath, JSON.stringify(summary, null, 2));
    
    // Print summary
    console.log('\n' + '='.repeat(60));
    console.log('📊 EXTRACTION SUMMARY');
    console.log('='.repeat(60));
    console.log(`Release ID: ${release.id}`);
    console.log(`Title: ${release.title}`);
    console.log(`Performer: ${release.primaryPerformer}`);
    console.log(`Audio Sources: ${release.audioSources.length}`);
    console.log(`Platforms: ${summary.platforms.join(', ')}`);
    console.log('\nExtracted Files:');
    
    for (const file of summary.extracted_files) {
      console.log(`  - ${file.platform}: ${file.file}`);
      console.log(`    Size: ${(file.size / 1024 / 1024).toFixed(2)} MB`);
      if (file.checksum) {
        console.log(`    SHA256: ${file.checksum.substring(0, 16)}...`);
      }
    }
    
    console.log(`\n✅ Complete release saved to: ${path.dirname(summaryPath)}`);
    console.log('='.repeat(60));
    
    return summary;
  }
}

// CLI Interface
async function main() {
  const args = process.argv.slice(2);
  
  if (args.length === 0 || args[0] === '--help') {
    console.log(`
Process Reddit URL - Complete GWASI Extraction Pipeline

Usage: node process-reddit-url.js <reddit_url> [options]

Arguments:
  reddit_url    Full Reddit post URL (e.g., https://www.reddit.com/r/gonewildaudio/comments/...)

Options:
  --output <dir>    Output directory (default: data)
  --no-cache        Disable caching
  --help            Show this help message

Example:
  node process-reddit-url.js "https://www.reddit.com/r/gonewildaudio/comments/1lxhwbd/f4m_catching_a_shy_touchstarved_girl_masturbating/"

Output Structure:
  data/
  ├── audio/          # Downloaded audio files by platform
  ├── enrichment/     # Reddit post analysis
  └── releases/       # Complete releases with all metadata
`);
    return 0;
  }
  
  const redditUrl = args[0];
  const options = {
    outputDir: 'data'
  };
  
  // Parse options
  for (let i = 1; i < args.length; i++) {
    if (args[i] === '--output' && args[i + 1]) {
      options.outputDir = args[++i];
    } else if (args[i] === '--no-cache') {
      options.cacheEnabled = false;
    }
  }
  
  try {
    const processor = new RedditProcessor(options);
    const result = await processor.processRedditUrl(redditUrl);
    
    if (result.success) {
      console.log('\n✅ Processing completed successfully!');
      return 0;
    } else {
      console.error('\n❌ Processing failed');
      return 1;
    }
    
  } catch (error) {
    console.error('\n❌ Fatal error:', error.message);
    if (error.stack) {
      console.error(error.stack);
    }
    return 1;
  }
}

if (require.main === module) {
  main().then(code => process.exit(code));
}

module.exports = RedditProcessor;

File: aural/process_reddit_url.py
#!/usr/bin/env python3
"""
Process Reddit URL - Complete End-to-End Workflow

Takes a Reddit URL as input and:
1. Fetches the Reddit post
2. Analyzes it with LLM
3. Extracts all audio files
4. Creates a complete release

Usage:
    uv run python process_reddit_url.py <reddit_url> [--output <dir>] [--no-cache]
"""

import argparse
import json
import re
import shutil
from pathlib import Path

import praw
from analyze_reddit_post import EnhancedRedditPostAnalyzer
from dotenv import load_dotenv
from release_orchestrator import ReleaseOrchestrator


# Load environment variables
load_dotenv()


class RedditProcessor:
    """Process Reddit URLs through the complete extraction pipeline."""

    def __init__(self, config: dict | None = None):
        config = config or {}
        self.output_dir = Path(config.get("outputDir", "data"))
        self.temp_dir = self.output_dir / ".temp"
        self.cache_enabled = config.get("cacheEnabled", True)

        self.orchestrator = ReleaseOrchestrator(
            {
                "dataDir": str(self.output_dir),
                "cacheEnabled": self.cache_enabled,
                "validateExtractions": True,
            }
        )

        # Reddit API client (initialized lazily)
        self._reddit: praw.Reddit | None = None

    @property
    def reddit(self) -> praw.Reddit:
        """Lazy initialization of Reddit API client."""
        if self._reddit is None:
            import os

            client_id = os.getenv("REDDIT_CLIENT_ID")
            client_secret = os.getenv("REDDIT_CLIENT_SECRET")
            user_agent = os.getenv("REDDIT_USER_AGENT", "Aural/1.0")

            if not client_id or not client_secret:
                raise ValueError(
                    "Reddit API credentials not found. "
                    "Set REDDIT_CLIENT_ID and REDDIT_CLIENT_SECRET in .env"
                )

            self._reddit = praw.Reddit(
                client_id=client_id,
                client_secret=client_secret,
                user_agent=user_agent,
            )
        return self._reddit

    def process_reddit_url(self, reddit_url: str) -> dict:
        """Process a Reddit URL through the complete pipeline."""
        print(f"Processing Reddit URL: {reddit_url}")
        print("=" * 60)

        try:
            # Ensure temp directory exists
            self.temp_dir.mkdir(parents=True, exist_ok=True)

            # Step 1: Analyze Reddit post
            print("\nStep 1: Analyzing Reddit post...")
            analysis = self.analyze_reddit_post(reddit_url)
            print(f"Analysis complete: {analysis['metadata']['title']}")

            # Step 2: Extract Reddit post data
            print("\nStep 2: Extracting post metadata...")
            reddit_post = self.extract_reddit_data(analysis)
            audio_count = len(analysis.get("audio_versions") or [])
            print(f"Found {audio_count} audio version(s)")

            # Step 3: Process into release
            print("\nStep 3: Extracting audio files...")
            release = self.orchestrator.process_post(reddit_post, analysis)

            # Step 4: Generate summary
            print("\nStep 4: Generating summary...")
            summary = self.generate_summary(release, reddit_url)

            # Cleanup temp directory
            try:
                shutil.rmtree(self.temp_dir)
            except OSError:
                pass  # Ignore cleanup errors

            return {"success": True, "release": release, "summary": summary}

        except Exception as error:
            print(f"\nProcessing failed: {error}")
            raise

    def analyze_reddit_post(self, reddit_url: str) -> dict:
        """Analyze Reddit post using the LLM analyzer."""
        analysis_output = self.temp_dir / "analysis.json"

        try:
            # First, we need to fetch the Reddit post and save it for the analyzer
            post_data = self.fetch_reddit_post(reddit_url)

            if not post_data:
                raise ValueError("Failed to fetch Reddit post")

            # Save post data for the analyzer
            post_file = self.temp_dir / "post_data.json"
            post_file.write_text(
                json.dumps(post_data, indent=2, ensure_ascii=False), encoding="utf-8"
            )

            # Run the LLM analyzer
            analyzer = EnhancedRedditPostAnalyzer()
            analysis = analyzer.analyze_post(post_file)

            # Save analysis result
            analysis_output.write_text(
                json.dumps(analysis, indent=2, ensure_ascii=False), encoding="utf-8"
            )

            return analysis

        except Exception as error:
            # Fallback to basic analysis
            print(f"Could not run LLM analysis, using basic extraction: {error}")
            return self.basic_reddit_analysis(reddit_url)

    def fetch_reddit_post(self, reddit_url: str) -> dict | None:
        """Fetch Reddit post data using PRAW."""
        post_id = self.extract_post_id(reddit_url)
        if not post_id:
            return None

        try:
            submission = self.reddit.submission(id=post_id)
            # Trigger lazy loading
            _ = submission.title

            # Check for crosspost and resolve if needed
            selftext = submission.selftext
            if self.is_crosspost(submission):
                resolved = self.resolve_crosspost(submission)
                if resolved:
                    selftext = resolved.get("selftext", selftext)

            return {
                "post_id": submission.id,
                "username": str(submission.author)
                if submission.author
                else "[deleted]",
                "reddit_url": reddit_url,
                "date": submission.created_utc,
                "reddit_data": {
                    "post_id": submission.id,
                    "title": submission.title,
                    "selftext": selftext,
                    "author": str(submission.author)
                    if submission.author
                    else "[deleted]",
                    "created_utc": submission.created_utc,
                    "subreddit": str(submission.subreddit),
                    "link_flair_text": submission.link_flair_text,
                    "url": submission.url,
                    "permalink": f"https://www.reddit.com{submission.permalink}",
                },
            }
        except Exception as error:
            print(f"Failed to fetch Reddit post: {error}")
            return None

    def is_crosspost(self, submission) -> bool:
        """Detect if a submission is a crosspost that needs resolution."""
        selftext = getattr(submission, "selftext", "") or ""
        if selftext.strip() and selftext not in ("[deleted]", "[removed]"):
            return False

        # Check explicit crosspost indicators
        if not submission.is_self:
            url = getattr(submission, "url", "") or ""
            if ("/r/" in url or "/u_" in url) and "/comments/" in url:
                return True

        # Check domain
        domain = getattr(submission, "domain", "") or ""
        if domain == "reddit.com":
            return True

        # User profile crossposts
        if not submission.is_self and domain.startswith("self."):
            return True

        # Check crosspost_parent_list
        crosspost_parent_list = getattr(submission, "crosspost_parent_list", None)
        if crosspost_parent_list and len(crosspost_parent_list) > 0:
            return True

        return False

    def resolve_crosspost(self, submission) -> dict | None:
        """Resolve a crosspost to get the original post content."""
        # First, check crosspost_parent_list
        crosspost_parent_list = getattr(submission, "crosspost_parent_list", None)
        if crosspost_parent_list and len(crosspost_parent_list) > 0:
            parent = crosspost_parent_list[0]
            parent_selftext = parent.get("selftext", "")
            if parent_selftext and parent_selftext not in ("[deleted]", "[removed]"):
                print("  Using crosspost_parent_list data")
                return {
                    "selftext": parent_selftext,
                    "resolved_from": "crosspost_parent_list",
                    "original_post_id": parent.get("id"),
                    "original_author": parent.get("author"),
                }

        # Try to resolve via URL
        url = getattr(submission, "url", "") or ""
        if "/comments/" in url:
            target_post_id = self.extract_post_id(url)
            if target_post_id and target_post_id != submission.id:
                print(f"  Resolving crosspost: {url}")
                try:
                    original = self.reddit.submission(id=target_post_id)
                    _ = original.title  # Trigger lazy load
                    original_selftext = getattr(original, "selftext", "") or ""

                    if original_selftext and original_selftext not in (
                        "[deleted]",
                        "[removed]",
                    ):
                        return {
                            "selftext": original_selftext,
                            "resolved_from": url,
                            "original_post_id": original.id,
                            "original_author": str(original.author)
                            if original.author
                            else "[deleted]",
                        }
                except Exception as error:
                    print(f"  Could not resolve crosspost: {error}")

        return None

    def extract_post_id(self, reddit_url: str) -> str | None:
        """Extract post ID from Reddit URL."""
        match = re.search(r"comments/([a-z0-9]+)", reddit_url)
        return match.group(1) if match else None

    def basic_reddit_analysis(self, reddit_url: str) -> dict:
        """Basic Reddit analysis fallback when LLM analysis fails."""
        post_data = self.fetch_reddit_post(reddit_url)

        if not post_data:
            raise ValueError("Failed to fetch Reddit post")

        reddit_data = post_data.get("reddit_data", {})
        content = reddit_data.get("selftext", "")

        # Extract audio URLs
        audio_urls = self.extract_audio_urls(content)

        return {
            "metadata": {
                "post_id": reddit_data.get("post_id"),
                "title": reddit_data.get("title"),
                "username": reddit_data.get("author"),
                "created_utc": reddit_data.get("created_utc"),
                "url": reddit_url,
            },
            "content": {
                "title": reddit_data.get("title"),
                "selftext": content,
            },
            "audio_versions": (
                [
                    {
                        "version_name": "Main",
                        "urls": [
                            {"url": url, "platform": self.detect_platform(url)}
                            for url in audio_urls
                        ],
                    }
                ]
                if audio_urls
                else []
            ),
            "performers": {
                "primary": reddit_data.get("author"),
                "additional": [],
                "count": 1,
            },
            "tags": self.extract_tags((reddit_data.get("title") or "") + " " + content),
        }

    def extract_audio_urls(self, text: str) -> list[str]:
        """Extract audio URLs from text."""
        patterns = [
            r"https?://(?:www\.)?soundgasm\.net/u/[^\s\]\)]+",
            r"https?://(?:www\.)?whyp\.it/tracks/[^\s\]\)]+",
            r"https?://(?:www\.)?hotaudio\.net/u/[^\s\]\)]+",
        ]

        urls = []
        for pattern in patterns:
            matches = re.findall(pattern, text, re.IGNORECASE)
            # Clean up URLs that might have trailing parentheses
            for url in matches:
                if url.endswith(")") and "(" not in url:
                    url = url[:-1]
                urls.append(url)

        return list(set(urls))  # Remove duplicates

    def detect_platform(self, url: str) -> str:
        """Detect platform from URL."""
        if re.search(r"soundgasm\.net", url, re.IGNORECASE):
            return "Soundgasm"
        if re.search(r"whyp\.it", url, re.IGNORECASE):
            return "Whyp.it"
        if re.search(r"hotaudio\.net", url, re.IGNORECASE):
            return "HotAudio"
        return "Unknown"

    def extract_tags(self, text: str) -> list[str]:
        """Extract tags from text."""
        tags = []
        for match in re.finditer(r"\[([^\]]+)\]", text):
            tags.append(match.group(1))
        return tags

    def extract_reddit_data(self, analysis: dict) -> dict:
        """Convert analysis to Reddit post format for orchestrator."""
        metadata = analysis.get("metadata", {})
        return {
            "id": metadata.get("post_id"),
            "title": metadata.get("title"),
            "author": metadata.get("username"),
            "created_utc": metadata.get("created_utc"),
            "selftext": analysis.get("content", {}).get("selftext", ""),
            "url": metadata.get("url"),
        }

    def generate_summary(self, release, reddit_url: str) -> dict:
        """Generate summary report."""
        release_dict = release.to_dict()

        summary = {
            "reddit_url": reddit_url,
            "release_id": release_dict["id"],
            "title": release_dict["title"],
            "performer": release_dict["primaryPerformer"],
            "audio_sources": len(release_dict["audioSources"]),
            "platforms": list(
                {
                    s.get("metadata", {}).get("platform", {}).get("name")
                    for s in release_dict["audioSources"]
                    if s.get("metadata", {}).get("platform", {}).get("name")
                }
            ),
            "extracted_files": [],
        }

        # List extracted files
        for source in release_dict["audioSources"]:
            audio = source.get("audio", {})
            summary["extracted_files"].append(
                {
                    "platform": source.get("metadata", {})
                    .get("platform", {})
                    .get("name"),
                    "file": audio.get("filePath"),
                    "size": audio.get("fileSize"),
                    "checksum": audio.get("checksum", {}).get("sha256"),
                }
            )

        # Save summary
        llm_analysis = release_dict.get("enrichmentData", {}).get("llmAnalysis") or {}
        version_naming = llm_analysis.get("version_naming", {})

        if version_naming.get("release_directory"):
            release_dir = (
                self.output_dir
                / "releases"
                / release_dict["primaryPerformer"]
                / version_naming["release_directory"]
            )
        else:
            release_dir = (
                self.output_dir
                / "releases"
                / release_dict["primaryPerformer"]
                / release_dict["id"]
            )

        summary_path = release_dir / "summary.json"
        summary_path.write_text(
            json.dumps(summary, indent=2, ensure_ascii=False), encoding="utf-8"
        )

        # Print summary
        print("\n" + "=" * 60)
        print("EXTRACTION SUMMARY")
        print("=" * 60)
        print(f"Release ID: {release_dict['id']}")
        print(f"Title: {release_dict['title']}")
        print(f"Performer: {release_dict['primaryPerformer']}")
        print(f"Audio Sources: {len(release_dict['audioSources'])}")
        print(f"Platforms: {', '.join(summary['platforms'])}")
        print("\nExtracted Files:")

        for file_info in summary["extracted_files"]:
            print(f"  - {file_info['platform']}: {file_info['file']}")
            if file_info["size"]:
                size_mb = file_info["size"] / 1024 / 1024
                print(f"    Size: {size_mb:.2f} MB")
            if file_info["checksum"]:
                print(f"    SHA256: {file_info['checksum'][:16]}...")

        print(f"\nComplete release saved to: {release_dir}")
        print("=" * 60)

        return summary


def main():
    parser = argparse.ArgumentParser(
        description="Process Reddit URL - Complete Aural Extraction Pipeline",
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Arguments:
  reddit_url    Full Reddit post URL (e.g., https://www.reddit.com/r/gonewildaudio/comments/...)

Options:
  --output <dir>    Output directory (default: data)
  --no-cache        Disable caching

Example:
  uv run python process_reddit_url.py "https://www.reddit.com/r/gonewildaudio/comments/1abc123/..."

Output Structure:
  data/
  ├── audio/          # Downloaded audio files by platform
  ├── enrichment/     # Reddit post analysis
  └── releases/       # Complete releases with all metadata
""",
    )
    parser.add_argument("reddit_url", nargs="?", help="Reddit post URL to process")
    parser.add_argument(
        "--output",
        "-o",
        default="data",
        help="Output directory (default: data)",
    )
    parser.add_argument(
        "--no-cache",
        action="store_true",
        help="Disable caching",
    )

    args = parser.parse_args()

    if not args.reddit_url:
        parser.print_help()
        return 0

    try:
        processor = RedditProcessor(
            {
                "outputDir": args.output,
                "cacheEnabled": not args.no_cache,
            }
        )
        result = processor.process_reddit_url(args.reddit_url)

        if result["success"]:
            print("\nProcessing completed successfully!")
            return 0
        else:
            print("\nProcessing failed")
            return 1

    except Exception as error:
        print(f"\nFatal error: {error}")
        import traceback

        traceback.print_exc()
        return 1


if __name__ == "__main__":
    exit(main())
